### PR TITLE
feat: add aws cdk for retrieval api

### PR DESCRIPTION
**Pull request by kylehuntsman**
This is the beginnings to the AWS serverless retrieval API for collecting stats about retrievals from providers. Still a WIP. Not guaranteed to be runnable.

### CODE REVIEW

**Comment by kylehuntsman** [comment]
I just realized there was duplicate code in here at `retrieval-api/src/save-provider-retrieval/index.ts`. I was playing with structure. I think most of this can be paired down to just this file for the most part.

File: retrieval-api/src/models/provider-retrieval.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  endedAt?: string
+  errorMessage?: string
+  lastStage: string
+  peerId: string
+  retrievalId: string
+  sizeBytes?: number
+  startedAt: string
+}
+
+type ConstProviderRetrievalProps = {
```
Comment (rvagg) [inline retrieval-api/src/models/provider-retrieval.ts:98]: is the dupe necessary here? can't we just use `ProviderRetrievalProps` everywhere?

Comment (kylehuntsman) [reply retrieval-api/src/models/provider-retrieval.ts:98]: I'm using the conversion to a `Date` in the constructor as a validation step. In this case, the data to create the model will be a string, I validate by turning it into a date, and then it gets spit back out as a string.

Comment (kylehuntsman) [reply retrieval-api/src/models/provider-retrieval.ts:98]: I could probably do some type implementing/overriding though to make it less fragile.

File: retrieval-api/src/save-provider-retrieval/index.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    try {
+      await this.providerRetrievalRepo.save(request.providerRetrieval)
+    } catch(err) {
+      console.error("Could not save provider retrieval.");
+      throw err;
+    }
+  }
+}
+
+class SaveProviderRetrievalController {
```
Comment (hannahhoward) [inline retrieval-api/src/save-provider-retrieval/index.ts:23]: this appears to be a duplicate of the code in `retrieval-api/src/infra/controllers/save-provider-retrieval-controller.ts`. Am I missing something?

File: retrieval-api/src/save-provider-retrieval/index.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { PostgresProviderRetrievalRepo } from "../infra/repos/postgres-provider-retrieval-repo";
+import { ProviderRetrieval } from "../models/provider-retrieval";
+import { ValidationError } from "../models/validation-error";
+import { ProviderRetrievalRepo } from "../repos/provider-retrieval-repo";
+
+class SaveProviderRetrieval {
```
Comment (hannahhoward) [inline retrieval-api/src/save-provider-retrieval/index.ts:6]: this appears to duplicate the file immediately below this

**Review by hannahhoward — CHANGES_REQUESTED** [review]
I am somewhat ignorant as to how the actual Lambda setup works but in terms of baseline API design this seems sound.

The only thing I'd like to understand prior to merge is why there is duplicate code.

**Review by hannahhoward — APPROVED** [review]

File: retrieval-api/src/models/provider-retrieval.ts (hunk 1 above)
Comment (kylehuntsman) [reply retrieval-api/src/models/provider-retrieval.ts]: Wasn't able to do any inheriting or overwriting of properties, but I renamed the type to `ValidatedProviderRetrievalProps` to make it more clear what they are used for.

File: retrieval-api/src/models/event-name.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { ValidationError } from "./validation-error";
+
+export class EventName {
+  public static VALUES = ["connected", "query-asked", "proposed", "accepted", "first-byte-received", "phase-failed", "succeeded"];
```
Comment (hannahhoward) [inline retrieval-api/src/models/event-name.ts:4]: these are from the filclient list I assume? Better sync with @rvagg

File: retrieval-api/src/models/event-time.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { ValidationError } from "./validation-error";
+
+export class EventTime {
+  public static create(value: string): EventTime | ValidationError {
+    const date = new Date(value);
+    if(date.getTime() !== date.getTime()) {
```
Comment (hannahhoward) [inline retrieval-api/src/models/event-time.ts:6]: Is this the correct way validating a date works? it is some JS madness?

File: retrieval-api/src/models/retrieval-event.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+export class RetrievalEvent {
+  public static create(props: RetrievalEventProps): RetrievalEvent | ValidationError {
+    if(!props.retrievalId) return new ValidationError("retrievalId is required.");
+    if(!props.cid) return new ValidationError("cid is required.");
+    if(!props.phase) return new ValidationError("phase is required.");
+    if(!props.phaseStartTime) return new ValidationError("phaseStartTime is required.");
+    if(!props.eventName) return new ValidationError("eventName is required.");
+    if(!props.eventTime) return new ValidationError("eventTime is required.");
+    if(!props.eventDetails) return new ValidationError("eventDetails is required.");
```
Comment (hannahhoward) [inline retrieval-api/src/models/retrieval-event.ts:15]: I think there are cases where event details is blank

File: retrieval-api/src/infra/repos/postgres-retrieval-event-repo.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+export class PostgresRetrievalEventRepo implements RetrievalEventRepo {  
+  private pool: Pool;
+  
+  constructor(props?: PostgresRetrivalEventRepoProps) {
+    const poolProps = props?.poolConfig ?? undefined;
+    this.pool = new Pool(poolProps);
+  }
+
+  save(retrievalEvent: RetrievalEvent): Promise<void> {
+    const queryStr = `
```
Comment (hannahhoward) [inline retrieval-api/src/infra/repos/postgres-retrieval-event-repo.ts:15]: This SQL looks incomplete -- it appears to be capturing only the UUID & CID. But also there are three values inserted with only 2 columns. Is the schema defined in code somewhere or does it live directly in AWS?

**Review by hannahhoward — COMMENTED** [review]
Lots of things looking good. Love all the validations and verification of models.

Things to double check prior to merge:
- do even names sync up with https://github.com/application-research/autoretrieve/pull/97?
- I don't think event details should be required, based on that code
- check for any other fields that may be blank sometimes

Finally it looks like the DB code itself still needs some more work. I imagine it's hard to unit test so you'll be left working directly in AWS once it's up.

**Pull request merged**